python/dask_cudf/dask_cudf/sorting.py

```diff
@@ -160,16 +160,19 @@ def rearrange_by_hash(
 
     token = tokenize(df, columns, max_branch)
     shuffle_combine_token = "shuffle-combine-" + token
-    start = {
-        (shuffle_combine_token, 0, inp): (df._name, i)
-        if i < df.npartitions
-        else df._meta
-        for i, inp in enumerate(inputs)
-    }
+    shuffle_token = "shuffle-" + token
+
+    start = {}
+    end = {}
+
+    for i, inp in enumerate(inputs):
+        start[(shuffle_combine_token, 0, inp)] = (
+            (df._name, i) if i < df.npartitions else df._meta
+        )
+        end[(shuffle_token, i)] = (shuffle_combine_token, stages, inp)
```

Comment:
We can fuse this with the loop inside the list comprehension for `inputs` above.